libpod/container_top_linux.go

```diff
@@ -113,60 +118,146 @@ func (c *Container) GetContainerPidInformation(descriptors []string) ([]string,
+// That way we open everything on the host and do not depend on any container libs.
+func (c *Container) execPS(psArgs []string) ([]string, error) {
 	rPipe, wPipe, err := os.Pipe()
 	if err != nil {
 		return nil, err
 	}
 	defer wPipe.Close()
 	defer rPipe.Close()
 
-	rErrPipe, wErrPipe, err := os.Pipe()
-	if err != nil {
-		return nil, err
-	}
-	defer wErrPipe.Close()
-	defer rErrPipe.Close()
-
-	streams := new(define.AttachStreams)
-	streams.OutputStream = wPipe
-	streams.ErrorStream = wErrPipe
-	streams.AttachOutput = true
-	streams.AttachError = true
-
 	stdout := []string{}
 	go func() {
 		scanner := bufio.NewScanner(rPipe)
 		for scanner.Scan() {
 			stdout = append(stdout, scanner.Text())
 		}
 	}()
-	stderr := []string{}
-	go func() {
-		scanner := bufio.NewScanner(rErrPipe)
-		for scanner.Scan() {
-			stderr = append(stderr, scanner.Text())
-		}
-	}()
 
-	cmd := append([]string{"ps"}, args...)
-	config := new(ExecConfig)
-	config.Command = cmd
-	ec, err := c.Exec(config, streams, nil)
+	psPath, err := exec.LookPath("ps")
 	if err != nil {
 		return nil, err
-	} else if ec != 0 {
-		return nil, fmt.Errorf("runtime failed with exit status: %d and output: %s", ec, strings.Join(stderr, " "))
 	}
+	psFD, err := unix.Open(psPath, unix.O_PATH, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer unix.Close(psFD)
+	logrus.Debugf("Trying to execute %q from the host in the container", psPath)
+	psPath = fmt.Sprintf("/proc/self/fd/%d", psFD)
 
-	if logrus.GetLevel() >= logrus.DebugLevel {
-		// If we're running in debug mode or higher, we might want to have a
-		// look at stderr which includes debug logs from conmon.
-		for _, log := range stderr {
-			logrus.Debugf("%s", log)
+	args := append([]string{psPath}, psArgs...)
+
+	// Now get all shared libs from ps(1), if this fails it is likely a static
+	// binary so no further actin required.
+	cmd := exec.Command("ldd", psPath)
+	output, err := cmd.Output()
+	if err == nil {
+		logrus.Debug("ps is dynamically linked, open linker and shared libraries for it")
+		var preload []string
+		var linkerPath string
+		for _, line := range strings.Split(string(output), "\n") {
+			fields := strings.Fields(line)
+			if len(fields) > 3 {
+				// open the shared lib on the host as it will most likely not be in the container
+				logrus.Debugf("Open shared library for ps: %s", fields[2])
+				fd, err := unix.Open(fields[2], unix.O_PATH, 0)
+				if err == nil {
+					defer unix.Close(fd)
+					preload = append(preload, fmt.Sprintf("/proc/self/fd/%d", fd))
+				}
+			} else if len(fields) == 2 {
+				if path.IsAbs(fields[0]) {
+					// this should be the dynamic linker
+					logrus.Debugf("Using linker for ps: %s", fields[0])
+					linkFD, err := unix.Open(fields[0], unix.O_PATH|unix.O_CLOEXEC, 0)
+					if err != nil {
+						return nil, err
+					}
+					defer unix.Close(linkFD)
+					linkerPath = fmt.Sprintf("/proc/self/fd/%d", linkFD)
+				}
+			}
```

Comment:
Can we add an else case with a log, indicating that the output's being skipped?  That may come in handy in case we ever need to debug it.  Consider a switch.